feat: introduce DetectStuckOrDisabledSchedules trigger

archu270292 opened the pull request:
Closes https://github.com/kestra-io/plugin-kestra/issues/61.

Pending / Need Guidance

Would like guidance on resolving the Lombok builder issue or confirming if it’s safe to ignore for now.

Build succeeded with ./gradlew clean build -x test.
<img width="911" height="276" alt="image" src="https://github.com/user-attachments/assets/dba37567-1b89-43b7-b5fb-954e8f2cff09" />

Full test run (./gradlew clean test) fails due to Testcontainers trying to pull a private Kestra EE image (ghcr.io/kestra-io/kestra-ee:develop).

Attached below is the local Kestra Docker log showing that the new plugin (DetectStuckSchedules) executed successfully:
<img width="959" height="379" alt="image" src="https://github.com/user-attachments/assets/a8cdfdb3-d262-4ec2-84ae-e8f704abce07" />

archu270292 reviewed (commented):
### Summary
Added a new trigger `DetectStuckSchedules` to detect stuck or misconfigured schedule triggers.

### Changes done
- Implemented `DetectStuckSchedules` plugin  
- Verified successful execution in local Docker (`Type DetectStuckSchedules started` and completed with SUCCESS`)  
- Fixed META-INF registration (`io.kestra.plugin.kestra.triggers.DetectStuckSchedules`)  
- Added initial test class  

### Notes
- Lombok `builder()` warning appears in IDE but plugin compiles and runs successfully  
- Full `./gradlew clean build` fails due to Testcontainers pulling private image (`kestra-ee:develop`)  
- Below is the Docker log screenshot showing successful execution and UI:
<img width="925" height="435" alt="image" src="https://github.com/user-attachments/assets/e91ae2c6-e55a-4bc9-8edb-5ee27f85cab9" />
![Screenshot 2025-10-20 154506](https://github.com/user-attachments/assets/5933a6aa-d2c2-428d-8623-51bbdf2461e5)
<img width="951" height="380" alt="image" src="https://github.com/user-attachments/assets/66dae412-429a-404c-95bd-d155db098126" />

loicmathieu reviewed (changes requested):
I think there is a misconception here, what needs to be done is not to list flows and search for triggers with null cron tab, those would not be saved as we have validation in place.

What needs to be done is to mist all triggers for the trigger endpoint and found the one that should have been started but are not (based on their cron expression).

archu270292 commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckSchedules.java:
```diff
@@ -0,0 +1,146 @@
+
+        // Example placeholder: in a real scenario, fetch these dynamically
+        String namespace = runContext.flowInfo().namespace();
+        List<String> flowIds = List.of("flow1", "flow2");
```
Done — moved namespace and flowIds to task properties as suggested.

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,180 @@
+                                issues.add(name + " (⏰ Stuck: last execution was " + minutesLate + " min ago)");
+                                runContext.logger().warn("⚠️ Stuck: {} ({} minutes late)", name, minutesLate);
+                            } else {
+                                runContext.logger().info("✅ {} OK | last execution {} min ago", name, minutesLate);
```
Same here, remove logging

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,172 @@
+
+            List<TriggerControllerTriggers> results = triggersResponse.getResults();
+            int returned = results.size();
+            runContext.logger().info("📦 Trigger API returned {} triggers (page {})", returned, page);
```
```suggestion
```

Please avoid excessive logging

archu270292 replied:
Excessive logging has been removed — only one global INFO log remains (Detection complete: X issues found).
All detailed trigger information is now returned through structured output fields instead of logs.

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,172 @@
+                    continue;
+                }
+
+                if (lastExec != null) {
```
A stuck trigger is detected not based on a thresold and the lastExecution but based on the `nextExecutionDate` of the trigger. The thresold is here to avoid warning for stuch trigger too quickly as we want to allow for ex to warn only if the schedule is missing since 5mn.

archu270292 commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,172 @@
+        String namespaceToUse = this.namespace != null ? this.namespace : runContext.flowInfo().namespace();
+
+        runContext.logger().info(
+            "🔍 Detecting stuck or disabled schedule triggers (tenant='{}', namespace='{}', threshold={} min)",
```
All emojis have been removed from every log statement.
The logs now follow a clean, consistent format without any special characters.

archu270292 commented:
![output](https://github.com/user-attachments/assets/a5721735-9ecf-47da-a4d1-77f7de2cbfe4)

Removed emojis and excessive logs 

Added structured output for stuckTriggers, disabledTriggers, and totalFound 

Logic correctly uses nextExecutionDate with a threshold delay 

Tested in Kestra — verified both healthy and stuck trigger cases

archu270292 commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,172 @@
+            for (TriggerControllerTriggers trigger : results) {
+
+                if (trigger.getAbstractTrigger() == null ||
+                    !trigger.getAbstractTrigger().getType().contains("Schedule")) {
```
done

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,186 @@
+
+    @Schema(title = "Threshold in minutes to consider a schedule overdue.")
+    @Builder.Default
+    private Property<Integer> thresholdMinutes = Property.ofValue(5);
```
Please use a `Duration` instead with a default of one minute

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,186 @@
+        private Integer totalFound;
+
+        @Schema(title = "Readable summary for logging and UI")
+        private String summary;
```
Please remove this, it's not needed to add a summary, you already log a summary

archu270292 replied:
Removed the summary field as suggested

loicmathieu commented on src/main/resources/META-INF/services/io.kestra.core.plugins.Plugin:
Why do you add this?
Where do you read that you need to do this?

archu270292 commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,186 @@
+            List<TriggerControllerTriggers> results = triggersResponse.getResults();
+
+
+            if (results.isEmpty()) {
```
updated pagination logic to correctly exit when the last page is reached.
Verified in Kestra UI (both happy and stuck trigger cases) — no infinite loop observed

archu270292 commented:
Replaced thresholdMinutes with Duration (default: 1 minute).

Removed summary field and unnecessary META-INF service file.

Fixed pagination logic to prevent infinite loop.

Updated test to use builder-based initialization (removed reflection).

Verified plugin functionality manually in Kestra UI:

Happy case: No stuck triggers detected.

Stuck case: Correctly detected 1 stuck trigger after schedule delay.

Screenshots attached:

Happy case result — shows detection complete with 0 issues.
![Happycaselog](https://github.com/user-attachments/assets/dac42fca-4d13-48f7-a82e-cd905657525e)


![Happycaseoutput](https://github.com/user-attachments/assets/0a397dc9-1e8f-4b00-8bcf-fdb3482bcfc4)




Stuck case result — shows detection complete with 1 stuck trigger.

![detectstuckcaselog](https://github.com/user-attachments/assets/a859bf0e-0b14-4711-98c2-9650655ca239)
![detectstuckoutput](https://github.com/user-attachments/assets/1d09ba3a-6511-45a8-b4c9-b5be20efce57)

loicmathieu commented on src/main/java/io/kestra/plugin/kestra/triggers/DetectStuckOrDisabledSchedules.java:
```diff
@@ -0,0 +1,191 @@
+
+Duration thresholdDuration = runContext.render(this.threshold)
+            .as(Duration.class)
+            .orElse(Duration.ofMinutes(1));
```
```suggestion
        Duration thresholdDuration = runContext.render(this.threshold)
            .as(Duration.class)
            .orElse(Duration.ofMinutes(1));
```

loicmathieu commented on src/test/java/io/kestra/plugin/kestra/triggers/DetectStuckSchedulesTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertThat(output.getStuckTriggers(), notNullValue());
+        assertThat(output.getTotalFound(), greaterThanOrEqualTo(0));
+
+        System.out.println("Test completed successfully. Total issues found: " + output.getTotalFound());
```
```suggestion
```

Please never use System.out

loicmathieu commented on src/test/java/io/kestra/plugin/kestra/triggers/DetectStuckSchedulesTest.java:
Your tests didn't test anything as you didn't inject triggers!
Can you add multiple triggers using the Kestra Client, you can see how other test does it for ex: https://github.com/kestra-io/plugin-kestra/blob/main/src/test/java/io/kestra/plugin/namespaces/ListTest.java

Idealy you should add:
- at least one non disabled and non-stuck trigger
- at least one disabled trigger
- at least one stuck trigger

And assert exactly on the result

archu270292 replied:
Added test case for detecting stuck and disabled schedules using KestraClient.
Unable to execute locally due to container image pull issue (EE image), tried alternate approaches to override with OSS image.
Test logic validated — ready for review.

archu270292 commented:
Covered enabled, disabled, and stuck trigger scenarios using KestraClient. Unable to execute locally due to container image pull issue (EE image), tried alternate approaches to override with OSS image.

![dockerimageerror](https://github.com/user-attachments/assets/207594d4-0000-4c25-9683-30ea7d78063d)


Cleaned up test code (removed System.out, proper assertions used).

loicmathieu commented on src/test/java/io/kestra/plugin/kestra/triggers/DetectStuckSchedulesTest.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    protected String kestraImage = "ghcr.io/kestra-io/kestra:develop";
+
```
```suggestion
```

Should not be needed

archu270292 replied:
removed. I’m was unable to build locally because the test fails to pull the ghcr.io/kestra-io/kestra-ee:develop image due to a permission denied error.
The System.setProperty("KESRA_IMAGE", "ghcr.io/kestra-io/kestra:develop") I had added earlier was only to test if  bypass this locally.
